mrmustard/lab_dev/states/states.py

```diff
@@ -0,0 +1,179 @@
+
+        >>> state = Coherent(modes=[0, 1, 2], x=[0.3, 0.4, 0.5], y=0.2)
+        >>> assert state.modes == [0, 1, 2]
+
+    Args:
+        modes: The modes of the coherent state.
+        x: The `x` displacement of the coherent state.
+        y: The `y` displacement of the coherent state.
+        x_trainable: Whether the `x` displacement is trainable.
+        y_trainable: Whether the `y` displacement is trainable.
+        x_bounds: The bounds of the `x` displacement.
+        y_bounds: The bounds of the `y` displacement.
+
+    .. details::
+
+        For any :math:`\bar{\alpha} = \bar{x} + i\bar{y}` of length :math:`N`, the :math:`N`-mode
+        coherent state displaced :math:`N`-mode vacuum state is defined by
+
+        .. math::
+            V = \frac{\hbar}{2}I_N \text{and } r = \sqrt{2\hbar}[\text{Re}(\bar{\alpha}), \text{Im}(\bar{\alpha})].
+
+        Its ``(A,b,c)`` triple is given by
+
+        .. math::
+            A = O_{N\text{x}N}\text{, }b=\bar{\alpha}\text{, and }c=\text{exp}\big(-|\bar{\alpha}^2|/2\big).
+    """
+
+    def __init__(
+        self,
+        modes: Iterable[int],
+        x: Union[float, Iterable[float]] = 0.0,
+        y: Union[float, Iterable[float]] = 0.0,
+        x_trainable: bool = False,
+        y_trainable: bool = False,
+        x_bounds: Tuple[Optional[float], Optional[float]] = (None, None),
+        y_bounds: Tuple[Optional[float], Optional[float]] = (None, None),
+    ):
+        super().__init__("Coherent", modes=modes)
+        self._add_parameter(make_parameter(x_trainable, x, "x", x_bounds))
+        self._add_parameter(make_parameter(y_trainable, y, "y", y_bounds))
+
+    @property
+    def representation(self) -> Bargmann:
+        n_modes = len(self.modes)
+        xs, ys = list(reshape_params(n_modes, x=self.x.value, y=self.y.value))
+        return Bargmann(*triples.coherent_state_Abc(xs, ys))
+
+
+class Number(Ket):
+    r"""The `N`-mode number state.
+
+    .. code-block::
+
+        >>> from mrmustard.lab_dev import Number
+
+        >>> state = Number(modes=[0, 1], n=[10, 20], cutoff=20)
+        >>> assert state.modes == [0, 1]
+
+    Args:
+        modes: The modes of the number state.
+        n: The number of photons in each mode.
+        cutoff: The cutoff. If ``None``, it defaults to the value of ``AUTOCUTOFF_MAX_CUTOFF`` in the settings.
+
+    .. details::
+
+        For any :math:`\bar{n} = (n_1,\:\ldots,\:n_N)`, the :math:`N`-mode number state is defined
+        by
+
+        .. math::
+            \ket{\bar{n}} = \ket{n_1}\otimes\ldots\otimes\ket{n_N}\:,
+
+        where :math:`\ket{n_j}` is the eigenstate of the number operator on mode `j` with eigenvalue
+        :math:`n_j`.
+
+    """
+
+    def __init__(
+        self, modes: Iterable[int], n: Union[int, Iterable[int]], cutoff: Optional[int] = None
+    ) -> None:
+        super().__init__("N", modes=modes)
+        self._add_parameter(make_parameter(False, n, "n", (None, None)))
+        self._add_parameter(
```

Comment:
would this parameter means "trainable" ones? cutoffs is not a trainable parameter ?

Reply:
They would not be, because of the `False`.
But I removed the `_add_parameter` to make it more explicit, and added properties and errors as needed